src/Compiler/Optimize/Optimizer.fs

```diff
@@ -230,20 +230,23 @@ type Summary<'Info> =
 // Note, this is a different notion of "size" to the one used for inlining heuristics
 //------------------------------------------------------------------------- 
 
-let rec SizeOfValueInfos (arr:_[]) =
-    if arr.Length <= 0 then 0 else max 0 (SizeOfValueInfo arr[0])
-
-and SizeOfValueInfo x =
-    match x with
-    | SizeValue (vdepth, _v) -> vdepth // terminate recursion at CACHED size nodes
-    | ConstValue (_x, _) -> 1
-    | UnknownValue -> 1
-    | ValValue (_vr, vinfo) -> SizeOfValueInfo vinfo + 1
-    | TupleValue vinfos
-    | RecdValue (_, vinfos)
-    | UnionCaseValue (_, vinfos) -> 1 + SizeOfValueInfos vinfos
-    | CurriedLambdaValue _ -> 1
-    | ConstExprValue (_size, _) -> 1
+let SizeOfValueInfo x =    
+    let rec SizeOfValueInfosHelper (arr:_[]) acc =
+        if arr.Length = 0 then acc else SizeOfValueInfoHelper arr[0] acc
+
+    and SizeOfValueInfoHelper x acc =
+        match x with
+        | SizeValue (vdepth, _v) -> max 0 (acc + vdepth) // terminate recursion at CACHED size nodes
+        | ConstValue (_x, _) -> acc + 1
+        | UnknownValue ->  acc + 1
+        | ValValue (_vr, vinfo) -> SizeOfValueInfoHelper vinfo (acc + 1)
+        | TupleValue vinfos
+        | RecdValue (_, vinfos)
+        | UnionCaseValue (_, vinfos) -> SizeOfValueInfosHelper vinfos (acc + 1)
+        | CurriedLambdaValue _ -> acc + 1
+        | ConstExprValue (_size, _) -> acc + 1
+
+    SizeOfValueInfoHelper x 0
```

Comment:
This could be simplified to a single local function by pattern-matching on empty arrays:

```suggestion
let SizeOfValueInfo valueInfo =
    let rec loop acc valueInfo =
        match valueInfo with
        | SizeValue (vdepth, _v) -> acc + max vdepth 0 // terminate recursion at CACHED size nodes
        | ValValue (_vr, vinfo) -> loop (acc + 1) vinfo
        | TupleValue [||]
        | RecdValue (_, [||])
        | UnionCaseValue (_, [||]) -> acc
        | TupleValue vinfos
        | RecdValue (_, vinfos)
        | UnionCaseValue (_, vinfos) -> loop (acc + 1) vinfos[0]
        | CurriedLambdaValue _
        | ConstExprValue _
        | ConstValue _
        | UnknownValue -> acc + 1

    loop 0 valueInfo
```

Also, it does seem odd that `vdepth` in `SizeValue (vdepth, _)` should ever be negative, but if it somehow were, then I think we'd want to do `acc + max vdepth 0` instead of `max 0 (acc + vdepth)`: the former behaves the same as the existing code, while the latter would produce a different result altogether if `acc` were a positive number with an absolute value greater than `vdepth`'s.

Reply:
Hey @brianrourkeboll, thanks a lot for the review and catching the issue in the `SizeValue` case.
I pushed your rewrite with a small change for matching the empty array.
Doing the empty check via length doesn't produce a null check like matching on `[| |]`
I updated the benchmark numbers.

Reply:
Ha, good catch—I don't know that I've ever consciously relied on a match on `[||]` doing a null check before, although I understand why it would do one.